### PR TITLE
improved read access for SDM630

### PR DESCRIPTION
**Pull request by f-reiling**
read access improved by reading several registers in one cycle -> higher update rate of meter values

## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ ] The code change is tested and works with Tasmota core ESP8266 V.2.7.7
  - [x] The code change is tested and works with Tasmota core ESP32 V.3.0.4
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by arendst** [comment]
And the impact on compiled file size is?

File: tasmota/tasmota_xnrg_energy/xnrg_10_sdm630.ino [hunk 1]
```diff
@@ -95,117 +92,72 @@ void SDM630Every250ms(void)
...
 
-        case 22:
-//          Energy->import_active[0] = value;
           EnergyUpdateTotal();
           break;
       }
 
       Sdm630.read_state++;
-      if (sizeof(sdm630_start_addresses)/2 == Sdm630.read_state) {
+      if ( (sizeof(sdm630ReqConf)/sizeof(sdm630ReqConf[0])) == Sdm630.read_state) {
```
Comment (eku) [inline tasmota/tasmota_xnrg_energy/xnrg_10_sdm630.ino:152]: Just wondering why we do not habe a macro `NELEM` for this. Saw this on many POSIX systems.

Comment (arendst) [reply tasmota/tasmota_xnrg_energy/xnrg_10_sdm630.ino:152]: Well I have a macro called ``nitems`` which does just that ;-)

**Comment by f-reiling** [comment]
> And the impact on compiled file size is?

compared to current implementation +1.6k

**Comment by eku** [comment]
> read access improved by reading several registers in one cycle -> higher update rate of meter values

What real-life problem is to be solved here?

**Comment by f-reiling** [comment]
> > read access improved by reading several registers in one cycle -> higher update rate of meter values
> 
> What real-life problem is to be solved here?

It depends on the application. 
* for some use-cases it might be sufficient to have an update rate of approx. 5s per value.
* In case of a fast power controller (as i am currently working on), it is better to have a higher update rate
* as not all values are read in same cycle, the data is partly inconsistent (e.g. U*I != P) -> new implementation ensures consistent data

**Comment by eku** [comment]
> > And the impact on compiled file size is?
> 
> compared to current implementation +1.6k

You might want to define unions containing both the floats and the byte array for several values togetherand iterate over it. Should save you the repeated calls to `convBufToFloat`.

**Comment by f-reiling** [comment]
this will not work on esp32 platform as modbus transmits MSB first. therefore the order of bytes needs to be changed

**Comment by eku** [comment]
> this will not work on esp32 platform as modbus transmits MSB first. therefore the order of bytes needs to be changed

My suggestion concerned the list of elements and not their value. I realise that the byte order has to be changed.

**Comment by f-reiling** [comment]
i added a define to select between both implementations, default is the old one.
i rechecked impact on code/RAM. it seems that the addional code usage is approx. 0k1

**Comment by arendst** [comment]
Thx.

**Pull request merged**